
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
        <link rel="prev" href="../Preparation/">
      
      
        <link rel="next" href="../Covplot/">
      
      
      <link rel="icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.6.20">
    
    
      
        <title>Virus scan - </title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.e53b48f4.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    <body dir="ltr">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#running-a-virus-scan-wolfpack" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="" class="md-header__button md-logo" aria-label="" data-md-component="logo">
      
  <img src="../assets/logo.svg" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Virus scan
            
          </span>
        </div>
      </div>
    </div>
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



  

<nav class="md-nav md-nav--primary md-nav--integrated" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="" class="md-nav__button md-logo" aria-label="" data-md-component="logo">
      
  <img src="../assets/logo.svg" alt="logo">

    </a>
    
  </label>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Home
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../Installation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Installation
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../Preparation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Preparation
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  
  <span class="md-ellipsis">
    Virus scan
    
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    Virus scan
    
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#wolfpack" class="md-nav__link">
    <span class="md-ellipsis">
      Wolfpack
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Wolfpack">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#additional-files" class="md-nav__link">
    <span class="md-ellipsis">
      Additional files
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#hot-run" class="md-nav__link">
    <span class="md-ellipsis">
      Hot run
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../Covplot/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Coverage plots
    
  </span>
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  



<h1 id="running-a-virus-scan-wolfpack">Running a virus scan: <code>wolfpack</code></h1>
<h2 id="wolfpack">Wolfpack</h2>
<p>This can be run on a single file or on a directory. It will try to guess from
the naming scheme if it is a Miseq output directory (<em>i.e.</em> with
<code>Data/Intensities/BaseCalls/</code> structure) and analyze all fastq files in there.
The extension must be <code>.fastq</code> or <code>.fastq.gz</code>. It will then run a filtering
step based on quality, length and entropy (in short: reads with a lot of
repeats will be discarded), followed by a decontamination step where reads
of human/bacterial/bovine/fungal origin will be discarded. Finally, remaining
reads are <em>blasted</em> against the viral database. The list of organisms with the
count of reads is in files <code>orgs_list.csv</code> in the output directory
(naming is <code>virmet_output_...</code>). For example, if we have a directory named
<code>exp_01</code> with files</p>
<pre><code>exp_01/AR-1_S1_L001_R1_001.fastq.gz
exp_01/AR-2_S2_L001_R1_001.fastq.gz
exp_01/AR-3_S3_L001_R1_001.fastq.gz
exp_01/AR-4_S4_L001_R1_001.fastq.gz
</code></pre>
<p>we could run</p>
<pre><code>virmet wolfpack --run exp_01
</code></pre>
<p>and, after some time, find the results in <code>virmet_output_exp01</code>. Many files are
present, the most important ones being <code>orgs_list.csv</code> and <code>stats.tsv</code>. The
first lists the viral organisms found with a count of reads that could be
matched to them.</p>
<pre><code>[user@host test_virmet]$ cat virmet_output_exp_01/AR-1_S1/orgs_list.tsv
organism    reads
Human adenovirus 7  126
Human poliovirus 1 strain Sabin 45
Human poliovirus 1 Mahoney  29
Human adenovirus 3+11p  19
Human adenovirus 16 1
</code></pre>
<p>The second file is a summary of all reads analyzed for this sample and how many
were passing a specific step of the pipeline or matching a specific database.</p>
<pre><code>[user@host test_virmet]$ cat virmet_output_exp01/AR-1_S1/stats.tsv
raw_reads       6250
trimmed_too_short       462
low_entropy     1905
low_quality     0
passing_filter  3883
matching_humanGRCh38    3463
matching_bact1  0
matching_bact2  0
matching_bact3  0
matching_fungi1 0
matching_bt_ref 0
reads_to_blast  420
viral_reads     257
undetermined_reads      163
</code></pre>
<h3 id="additional-files">Additional files</h3>
<p>At the end of a run a directory for each sample (fastq file analyzed) is
created containing the following files:</p>
<pre><code>good_humanGRCh38_bact1_bact2_bact3_fungi1_bt_ref.cram
good_humanGRCh38_bact1_bact2_bact3_fungi1_bt_ref.err
...
good_humanGRCh38_bact1.cram
good_humanGRCh38_bact1.err
good_humanGRCh38.cram
good_humanGRCh38.err

orgs_list.tsv
prinseq.err
prinseq.log
stats.tsv
undetermined_reads.fastq.gz
unique.tsv.gz
viral_reads.fastq.gz
</code></pre>
<p>Files <code>orgs_list.tsv</code> and <code>stats.tsv</code> report the main output of the tool as
reported above, while <code>unique.tsv.gz</code> reports blast hits to viral database.</p>
<p>As the names say, <code>viral_reads.fastq.gz</code> and <code>undetermined_reads.fastq.gz</code>
contain, respectively, reads identified as of viral origin and reads not
matching any of the considered genomes.</p>
<p><code>prinseq.err</code> and <code>prinseq.log</code> are, respectively, the standard error and log
file of prinseq, used to filter reads. By inspecting this log file, VirMet
determines how many reads were discarded because of low entropy or low quality.</p>
<p>In the <em>decontamination</em> step, reads are aligned against human genome first,
those matching are discarded while those not matching are aligned against
the first set of bacterial genomes, and so on.
File <code>good_humanGRCh38.cram</code> is the alignment of high quality reads (good) to
human genome, saved in CRAM format. File <code>good_humanGRCh38_bact1.cram</code> contains
the alignment to bacterial genomes in set bact1 of high quality reads (good) minus
those that were identified as matching human genome, and so on. File ending in
<code>err</code> contain the standard error of the conversion bam -&gt; cram.</p>
<h3 id="hot-run">Hot run</h3>
<p>A virus scan on a full MiSeq run typically lasts a few hours, many of which are spent
in the decontamination phase. Sometimes, after a run is completed, we would like to
run it again with a new viral database. In these cases, <code>wolfpack</code> would run skipping
the previous phases to save time. It relies on the presence of intermediate files that,
if present, signals the pipeline that a specific step must be skipped.</p>
<p>These are the rules (must be intended for each sample):</p>
<ul>
<li>if both <code>prinseq.log</code> and <code>prinseq.err</code> exist, skip quality filtering,</li>
<li>if <code>good_humanGRCh38.err</code> exists, skip the human reads decontamination,</li>
<li>if <code>good_humanGRCh38_bact1.err</code> exists, skip the decontamination against first bacterial
  database, and so on.</li>
</ul>
<p>Blasting against viral database will always be performed. If both <code>viral_reads.fastq.gz</code>
and <code>undetermined_reads.fastq.gz</code> exist, their content will be copied into a file,
they will be removed, and this new file will be blasted against the viral database.</p>
<p>In short, if we change the viral database after a run has already been analyzed, simply
running <code>virmet wolfpack</code> again will skip the quality filtering and go straight to
blast against viral database.</p>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      <script id="__config" type="application/json">{"base": "..", "features": ["toc.integrate"], "search": "../assets/javascripts/workers/search.973d3a69.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../assets/javascripts/bundle.f55a23d4.min.js"></script>
      
    
  </body>
</html>