
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
        <link rel="prev" href="../Preparation/">
      
      
        <link rel="next" href="../Covplot/">
      
      
      <link rel="icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.6.20">
    
    
      
        <title>Virus scan - </title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.e53b48f4.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    <body dir="ltr">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#running-a-virus-scan-wolfpack" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="" class="md-header__button md-logo" aria-label="" data-md-component="logo">
      
  <img src="../assets/logo.svg" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Virus scan
            
          </span>
        </div>
      </div>
    </div>
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



  

<nav class="md-nav md-nav--primary md-nav--integrated" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="" class="md-nav__button md-logo" aria-label="" data-md-component="logo">
      
  <img src="../assets/logo.svg" alt="logo">

    </a>
    
  </label>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Home
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../Installation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Installation
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../Preparation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Preparation
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  
  <span class="md-ellipsis">
    Virus scan
    
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    Virus scan
    
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#wolfpack" class="md-nav__link">
    <span class="md-ellipsis">
      Wolfpack
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#outputs" class="md-nav__link">
    <span class="md-ellipsis">
      Outputs
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#hot-run" class="md-nav__link">
    <span class="md-ellipsis">
      Hot run
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../Covplot/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Coverage plots
    
  </span>
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  



<h1 id="running-a-virus-scan-wolfpack">Running a virus scan: <code>wolfpack</code></h1>
<h2 id="wolfpack">Wolfpack</h2>
<p>Wolfpack is the main subcommand of VirMet and it allows to perform
mNGS analyses of the given samples (virus scan).</p>
<p>Information on how to use it can be obtained with <code>-h</code>:</p>
<div class="highlight"><pre><span></span><code>virmet wolfpack -h
usage: virmet wolfpack [options]

Options:
  -h, --help       show this help message and exit
  --run RUN        Miseq run directory
  --file FILE      single fastq file
  --dbdir [DBDIR]  path to find and use the Virmet database. Default: /data/virmet_databases/
  --nocovplot      do not make the covplots. Default: make them
  --noctrls        do not analyse ntc- or Undetermined samples. Default: analyse them
</code></pre></div>
<p>Wolfpack can be run on a single file (<code>--file</code>) or on a directory (<code>--run</code>).
If a directory is provided, it analyses all the FASTQ files in there.
Please, note that Wolfpack works if the structure of the <code>--run</code> directory is
any of these 2:</p>
<div class="highlight"><pre><span></span><code># Structure 1
run_directory/
├── FASTQ_file_1
├── FASTQ_file_2
├── ...
└── FASTQ_file_n

# Structure 2
run_directory/
└── Data/
    └── Intensities/
        └── BaseCalls/
            ├── FASTQ_file_1
            ├── FASTQ_file_2
            ├── ...
            └── FASTQ_file_n
</code></pre></div>
<p>Importantly, the extension of the FASTQ files must be
<code>.fastq</code> or <code>.fastq.gz</code>.</p>
<p>To run Wolfpack, users have the option to specify <code>--dbdir</code>, which is
the path to the VirMet database and must include the viral, human, bovine, 
bacterial and fungal (bact_fungi) databases.
See how to <a href="../Preparation/">prepare for a VirMet run</a> before running
Wolfpack.
If <code>-dbdir</code> is not specified, VirMet will assume that it is located in
<code>/data/virmet_databases/</code>.</p>
<p>Users also have the option to run the mNGS analyses on all the FASTQ files
(the default behaviour), which usually include negative controls and 
undertermined reads, or to skip those with <code>--noctrls</code>.
In this case, all files named <code>ntc-*</code> or <code>Undetermined*</code> will be ignored.</p>
<p>To carry out the virus scan, Wolfpack will first run a filtering step (QC)
based on quality, length and entropy of the given reads (named quality-control).
Then, it will perform a decontamination step, where reads of human, bovine,
bacterial and fungal origin will be discarded.
Finally, remaining reads will be <em>blasted</em> against the viral database and
the viruses present in the mNGS samples will be identified.</p>
<p>Before providing the results, Wolfpack will (by default) automatically check
which of the identified viral organisms fulfill these criteria:</p>
<ul>
<li>≥3 sequencing reads match this specific viral organism.</li>
<li>The weighted mean coverage score for this viral organism is ≥10%.</li>
<li>The viral organism is not a phage nor a non-human virus (host is human).</li>
</ul>
<p>For all the viral organisms meeting these criteria, Wolfpack will generate
<a href="../Covplot/">Coverage plots</a>, where users can see what fraction of the genome is covered
by the sequenncing reads. This will provide them with further evidence
supporting (or not) the presence of the organism in the sample.</p>
<p>If users want to avoid this last step and are not interested in getting the
coverage plots, they can disable such option with <code>--nocovplot</code>.</p>
<h2 id="outputs">Outputs</h2>
<p>Virmet Wolfpack provides several outputs that can be found in the
output folder <code>virmet_output_RUN_NAME</code>, located at the current working directory.</p>
<p>For example, if users have a directory named <code>Exp01</code> with files:</p>
<div class="highlight"><pre><span></span><code>exp_01/ABC-DNA_S9.fastq.gz
exp_01/DEF-RNA_S11.fastq.gz
</code></pre></div>
<p>and run <code>virmet wolfpack --run Exp01</code>, their results will
appear into <code>virmet_output_Exp01</code>.</p>
<p>The most important output files are:</p>
<ul>
<li><strong>Orgs_species_found.csv</strong>: table showing the viral organisms identified per
FASTQ file as well as the read counts matching each organism.
It looks as follows:</li>
</ul>
<div class="highlight"><pre><span></span><code>species                accn        reads  stitle               ssciname               covered_region  seq_len  sample       run
Tunavirus T1           MK213796.1  455    Phage T1             Escherichia phage T1   30152           48836    ABC-DNA_S9   Exp01
Varicellovirus a5      KY559403.2  79     Alphaherpesvirus 5   Alphaherpesvirus 5     321             137741   ABC-DNA_S9   Exp01
Lentivirus humimdef1   MT222957.1  13     HIV-1 isolate X      HIV-1                  141             9372     DEF-RNA_S11  Exp01
Emesvirus zinderi      EF204940.1  6265   Phage MS2 isolate Y  Escherichia phage MS2  2836            3569     DEF-RNA_S11  Exp01
</code></pre></div>
<ul>
<li><strong>Run_reads_summary.tsv</strong>: summary of all reads analyzed per sample, showing the
number of reads passing each step of the pipeline (e.g., QC, decontamination),
and the number of reads matching the human, bovine, bacterial, fungal and
viral database.
It looks as follows:</li>
</ul>
<div class="highlight"><pre><span></span><code>category              reads     sample       run
raw_reads             2656448   ABC-DNA_S9   Exp01
trimmed_too_short     46        ABC-DNA_S9   Exp01
low_entropy           25        ABC-DNA_S9   Exp01
low_quality           1081      ABC-DNA_S9   Exp01
passing_filter        2655296   ABC-DNA_S9   Exp01
matching_humanGRCh38  339020    ABC-DNA_S9   Exp01
matching_bt_ref       1319509   ABC-DNA_S9   Exp01
matching_bacteria     46024     ABC-DNA_S9   Exp01
matching_fungi        10        ABC-DNA_S9   Exp01
matching_other_cells  7403      ABC-DNA_S9   Exp01
reads_to_blast        943330    ABC-DNA_S9   Exp01
viral_reads           70579     ABC-DNA_S9   Exp01
undetermined_reads    872751    ABC-DNA_S9   Exp01
raw_reads             2202164   DEF-RNA_S11  Exp01
trimmed_too_short     259441    DEF-RNA_S11  Exp01
low_entropy           18        DEF-RNA_S11  Exp01
low_quality           1277      DEF-RNA_S11  Exp01
passing_filter        1941428   DEF-RNA_S11  Exp01
matching_humanGRCh38  526874    DEF-RNA_S11  Exp01
matching_bt_ref       819466    DEF-RNA_S11  Exp01
matching_bacteria     8721      DEF-RNA_S11  Exp01
matching_fungi        21        DEF-RNA_S11  Exp01
matching_other_cells  299       DEF-RNA_S11  Exp01
reads_to_blast        586047    DEF-RNA_S11  Exp01
viral_reads           168247    DEF-RNA_S11  Exp01
undetermined_reads    417800    DEF-RNA_S11  Exp01
</code></pre></div>
<p>In addition to these main outputs, Wolfpack creates inside
<code>virmet_output_RUN_NAME</code> a subdirectory for each sample (FASTQ file).
There, users can find these additional files: </p>
<ul>
<li>Unique.tsv.gz</li>
<li>Viral_reads.fastq.gz</li>
<li>Undetermined_reads.fastq.gz</li>
<li>Orgs_list.tsv</li>
<li>Stats.tsv</li>
<li>Fastp.html</li>
<li><strong>Folders named as viral organisms</strong></li>
<li>Other.err</li>
</ul>
<p>Files <code>orgs_list.tsv</code> and <code>stats.tsv</code> report the main output of the tool for
each sample. They are the same as <code>Orgs_species_found.csv</code> and <code>Run_reads_summary.tsv</code>,
respectively, but containing information only of one sample. Therefore, the
last two columns (<em>sample</em> and <em>run</em>) are missing.</p>
<p>The <code>unique.tsv.gz</code> reports all BLAST hits to the viral database. These are
not the same as in <code>orgs_list.tsv</code> because VirMet further filters the
BLAST hits (<code>unique.tsv.gz</code>) to ensure that only those with ≥75% pident and
≥75% qcov are ultimately considered viral reads for diagnosis. </p>
<p>As the names say, <code>viral_reads.fastq.gz</code> and <code>undetermined_reads.fastq.gz</code>
contain, respectively, reads identified as of viral origin and reads not
matching any of the considered genomes.</p>
<p>Finally, <code>fastp.html</code> shows the quality-control statistics and information
about the QC-filtering step.</p>
<p>In the <em>decontamination</em> step, reads are aligned against the human genome first,
those matching are discarded while those not matching are aligned against
the bovine genome, and so on. In each step, some files ending with <em>.err</em>
are generated, and they can be used for inspection, if needed. However, they
do not provide valuable information for the users (unless there is any error)
and can be removed if desired.</p>
<p>Besides all these files, Wolfpack automatically creates (unless disabled with
<code>--nocovplot</code>) a few folders with the names of viral organisms. Such folders
contain the coverage plots of the viral assignments, named <code>Viral_organism_coverage.pdf</code>.
It is recommended to manually have a look at them before a final viral diagnosis.</p>
<p>Overall, users can expect the following structure for the outputs:</p>
<div class="highlight"><pre><span></span><code>virmet_output_Exp01/
│
├── ABC-DNA_S9
│   ├── Virus_X/
│   │    └── Virus_X_coverage.pdf
│   ├── unique.tsv.gz, viral_reads.fastq.gz, undetermined_reads.fastq.gz
│   ├── orgs_list.tsv, stats.tsv, fastp.html
│   └── Others.err
│ 
├── DEF-RNA_S11
│   ├── Virus_Y/
│   │   └── Virus_Y_coverage.pdf
│   ├── Virus_Z/
│   │   └── Virus_Z_coverage.pdf
│   ├── unique.tsv.gz, viral_reads.fastq.gz, undetermined_reads.fastq.gz
│   ├── orgs_list.tsv, stats.tsv, fastp.html
│   └── Others.err
│
├── run_reads_summary.tsv
└── orgs_species_found.tsv
</code></pre></div>
<p>More information about the <a href="../Covplot/">Coverage plots</a> is provided in the corresponding
section.</p>
<h2 id="hot-run">Hot run</h2>
<p>A virus scan on a full MiSeq run typically lasts a few hours, most of which are spent
in the decontamination phase. Sometimes, after a run is completed, we would like to
run it again with a new viral database. In these cases, <code>wolfpack</code> would run skipping
the previous phases to save time. It relies on the presence of intermediate files that,
if present, signals the pipeline that a specific step must be skipped.</p>
<p>These are the rules (must be intended for each sample):</p>
<ul>
<li>If <code>fastp.html</code> exists, skip quality filtering.</li>
<li>If <code>good_humanGRCh38.err</code> exists, skip the human reads decontamination.</li>
<li>If <code>good_humanGRCh38_bt_ref.err</code> exists, skip the bovine reads decontamination.</li>
<li>if <code>good_humanGRCh38_bt_ref_bact_fungi.err.err</code> exists, skip the 
decontamination bacterial and fungal database.</li>
</ul>
<p>Blasting against viral database will always be performed. If both <code>viral_reads.fastq.gz</code>
and <code>undetermined_reads.fastq.gz</code> exist, their content will be copied into a file,
they will be removed, and this new file will be blasted against the viral database.</p>
<p>Therefore, if users change the viral database after a run has already been analyzed,
running <code>virmet wolfpack</code> again will skip the quality filtering and decontamination steps
and proceed directly with the viral assignment.</p>
<p><strong>Enjoy with your viral classification!</strong></p>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      <script id="__config" type="application/json">{"base": "..", "features": ["toc.integrate"], "search": "../assets/javascripts/workers/search.973d3a69.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../assets/javascripts/bundle.f55a23d4.min.js"></script>
      
    
  </body>
</html>